src/types.rs

```diff
@@ -627,24 +629,32 @@ enum AbsurdComparisonResult {
     InequalityImpossible,
 }
 
+enum Rel {
+    Lt,
+    Le,
+}
+
+// Put the expression in the form lhs < rhs or lhs <= rhs.
+fn normalize_comparison<'a>(op: BinOp_, lhs: &'a Expr, rhs: &'a Expr)
+                                -> Option<(Rel, &'a Expr, &'a Expr)> {
+    match op {
+        BiLt => Some((Rel::Lt, lhs, rhs)),
+        BiLe => Some((Rel::Le, lhs, rhs)),
+        BiGt => Some((Rel::Lt, rhs, lhs)),
+        BiGe => Some((Rel::Le, rhs, lhs)),
+        _ => None,
+    }
+}
+
 fn detect_absurd_comparison<'a>(cx: &LateContext, op: BinOp_, lhs: &'a Expr, rhs: &'a Expr)
                                 -> Option<(ExtremeExpr<'a>, AbsurdComparisonResult)> {
     use types::ExtremeType::*;
     use types::AbsurdComparisonResult::*;
     type Extr<'a> = ExtremeExpr<'a>;
 
-    // Put the expression in the form lhs < rhs or lhs <= rhs.
-    enum Rel {
-        Lt,
-        Le,
-    };
-    let (rel, normalized_lhs, normalized_rhs) = match op {
-        BiLt => (Rel::Lt, lhs, rhs),
-        BiLe => (Rel::Le, lhs, rhs),
-        BiGt => (Rel::Lt, rhs, lhs),
-        BiGe => (Rel::Le, rhs, lhs),
-        _ => return None,
-    };
+    let normalized = normalize_comparison(op, lhs, rhs);
+    if normalized.is_none() { return None; } // Could be an if let, but this prevents rightward drift
+    let (rel, normalized_lhs, normalized_rhs) = normalized.expect("Unreachable-- is none check above");
```

Comment:
Worth adding a lint for? I'm not sure exactly what the match should be. An `is_none()` followed by an `expect(...)` on identical variables?